packages/caver-klay/caver-klay-kct/kip7.js

```diff
@@ -0,0 +1,388 @@
+/*
+    Copyright 2020 The caver-js Authors
+    This file is part of the caver-js library.
+
+    The caver-js library is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    The caver-js library is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with the caver-js. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+const _ = require('lodash')
+const BigNumber = require('bignumber.js')
+
+const Contract = require('../caver-klay-contract')
+const { validateTokenInfoForDeploy, determineSendParams, kip7JsonInterface, kip7ByteCode, formatParamForUint256 } = require('./kctHelper')
+const { isAddress } = require('../../caver-utils')
+
+class KIP7 extends Contract {
+    /**
+     * deploy deploys KIP-7 token contract to Klaytn network.
+     * `const deployedContract = await caver.klay.KIP7.deploy({
+     *      name: 'TokenName',
+     *      symbol: 'TKN',
+     *      decimals: 18,
+     *      initialSupply: new BigNumber(1000000000000000000),
+     *  }, '0x{address in hex}')`
+     *
+     * @method deploy
+     * @param {Object} tokenInfo The object that defines the name, symbol, decimals, and initialSupply of the token to deploy.
+     * @param {String} privateKey The address of the account to deploy the KIP-7 token contract.
+     * @return {Object}
+     */
+    static deploy(tokenInfo, deployer) {
+        validateTokenInfoForDeploy(tokenInfo)
+
+        const { name, symbol, decimals, initialSupply } = tokenInfo
+        const kip7 = new KIP7()
+
+        return kip7
+            .deploy({
+                data: kip7ByteCode,
+                arguments: [name, symbol, decimals, initialSupply],
+            })
+            .send({ from: deployer, gas: 3500000, value: 0 })
+    }
+
+    constructor(tokenAddress, abi = kip7JsonInterface) {
+        if (tokenAddress) {
+            if (_.isString(tokenAddress)) {
+                if (!isAddress(tokenAddress)) throw new Error(`Invalid token address ${tokenAddress}`)
+            } else {
+                abi = tokenAddress
+                tokenAddress = undefined
```

Comment:
What make you generate this code? It seems that you receive abi as the first parameter while tokenAddress is not passed.

Reply:
It's written this way because both parameters received by the constructor are implemented as optional.

By default, KIP7 class uses abi implemented in caver-js, but it is implemented as optional to allow the token contract abi, an extension of kip7.

The extension part is not provided as a method of the class, but if it is not provided, the contract which implements the extension kip7 cannot use KIP7 class at all, so it is implemented as optional.

For this purpose, I wrote a phrase in Klaytn Docs that says that it will not work properly if you pass an abi that does not follow KIP7 as a parameter.

Reply:
Maybe I misunderstood something. Is it okay if we do not receive the contract address?
This code sets tokenAddress as `undefined`.

Reply:
tokenAddress is a parameter of type string, and abi is of type object.

The condition corresponding to else in the above code is when the tokenAddress parameter is not of type string. 
In this case, only abi, not tokenAddress, is passed to the constructor.
In this case, enter the parameter value passed by the actual user in abi variable, and create an instance with tokenAddress set to undefined.

If the instance is created like this, the KIP7 instance is created without the address of the contract.
Currently, the methods of KIP7 instance are all functions that execute smart contracts(except clone), but there are functions in the Contract class besides the function of actually calling contracts.

Reply:
So, you mean, if we want to use some auxiliary functions (not calling the smart contract) in `Contract`, we can create an instance like this? `Contract(abi)`

Reply:
Yes, If it is a Contract instance, we can create 'new caver.klay.Contract(abi)' like this.
And if it is a KIP7 instance, we can create 'new caver.klay.KIP7 ()' or 'new caver.klay.KIP (extendedABI)'.